src/cell_group.hpp

```diff
@@ -143,51 +143,61 @@ class cell_group {
     }
 
     void advance(time_type tfinal, time_type dt) {
-        while (lowered_.time()<tfinal) {
-            // take any pending samples
-            time_type cell_time = lowered_.time();
+        EXPECTS(lowered_.state_synchronized());
+
+        // Bin pending events and enqueue on lowered state.
+        time_type ev_min_time = lowered_.max_time(); // (but we're synchronized here)
+        while (auto ev = events_.pop_if_before(tfinal)) {
+            auto handle = get_target_handle(ev->target);
+            auto binned_ev_time = binner_.bin(ev->target.gid, ev->time, ev_min_time);
+            lowered_.add_event(binned_ev_time, handle, ev->weight);
+        }
+
+        lowered_.start_integration(tfinal, dt);
```

Comment:
Happy to change to `init_integration` or similar. It does, in some sense, get things going: 'prepare' would be a bit misleading.

Either way though, indeed once sampling is also performed at the lowered level there will be no need to have separate methods here.